webhooks/cluster.go

```diff
@@ -182,34 +182,34 @@ func (webhook *Cluster) validateTopology(ctx context.Context, old, new *clusterv
 		)
 	}
 
-	// MachineDeployment names must be unique.
-	if new.Spec.Topology.Workers != nil {
-		names := sets.String{}
-		for _, md := range new.Spec.Topology.Workers.MachineDeployments {
-			if names.Has(md.Name) {
-				allErrs = append(allErrs,
-					field.Invalid(
-						field.NewPath("spec", "topology", "workers", "machineDeployments"),
-						md,
-						fmt.Sprintf("MachineDeployment names should be unique. MachineDeployment with name %q is defined more than once.", md.Name),
-					),
-				)
-			}
-			names.Insert(md.Name)
-		}
+	// clusterClass must exist.
+	clusterClass := &clusterv1.ClusterClass{}
+	// Check to see if the ClusterClass referenced in the Cluster currently exists.
+	if err := webhook.Client.Get(ctx, client.ObjectKey{Namespace: new.Namespace, Name: new.Spec.Topology.Class}, clusterClass); err != nil {
+		allErrs = append(
+			allErrs, field.Invalid(
+				field.NewPath("spec", "topology", "class"),
+				new.Name,
+				"ClusterClass could not be found"))
+		return allErrs
 	}
 
+	allErrs = append(allErrs, check.MachineDeploymentClassesAreUnique(clusterClass)...)
+	allErrs = append(allErrs, check.MachineDeploymentClassesAreUniqueAndDefinedInClusterClass(new, clusterClass)...)
+
 	if old != nil { // On update
-		// Class could not be mutated.
-		if new.Spec.Topology.Class != old.Spec.Topology.Class {
+		// Topology or Class can not be removed.
+		if old.Spec.Topology == nil || len(old.Spec.Topology.Class) == 0 {
 			allErrs = append(
 				allErrs,
 				field.Invalid(
 					field.NewPath("spec", "topology", "class"),
 					new.Spec.Topology.Class,
-					"class cannot be changed",
+					"cannot be empty",
```

Comment:
I think this is the unmanaged => managed case, maybe:

`unmanaged Cluster cannot be converted to managed`

Reply:
I wonder if we validate the reverse somewhere (old has class, new has not)

Reply:
Yeah - that's what I'm intent on blocking here. We're not blocking the opposite right now - I think it's easy to take a cluster out of clusterClass management the way it's currently set up, but it would break reconciliation.

Reply:
I'll add a check to block this for now and create an issue for it. I'm not sure how much the managed -> unmanaged use case matters.

Reply:
In general I think longterm we should allow both cases in the webhook. Doesn't really mean we have to build something to support those use cases, but currently we force users who want to do this to work around the webhook.

Reply:
> Yeah - that's what I'm intent on blocking here. We're not blocking the opposite right now - I think it's easy to take a cluster out of clusterClass management the way it's currently set up, but it would break reconciliation.

I'm not sure it would break reconciliation. I think you can just remove the entire topology config from your Cluster and it should still continue to work. But I think it makes sense to at least validate both cases once manually.

Reply:
The cluster will continue to work - but you'll have to replace the yaml with the API server version I guess to make future changes to the cluster, not an ideal UX, but it's unsupported right now.

Reply:
>In general I think longterm we should allow both cases in the webhook. Doesn't really mean we have to build something to support those use cases, but currently we force users who want to do this to work around the webhook.